### ISSUE TITLE
Suggest search engine selection based on typed text

### ISSUE THREAD

**Issue by brampitoyo** [mozilla-mobile/fenix#16044]
### What is the user problem or growth opportunity you want to see solved?

- Many people entered keywords into the address bar and waited for auto-suggest options before making a choice
- Many people entered a search engine name into the address bar and manually go there, but they ultimately want to perform a one-time search using that search engine

By suggesting one-time search engine selection in the address bar results, we can help people perform their search faster.

### Behaviour

- Search engine suggestion appears first on the list, ahead of the first (verbatim) search suggestion
- Search engine suggestion only appears when autocomplete matches a search engine name or URL
    - Typing “t” autocompletes to “twitter.com”
    - When this happens, show a suggestion labelled “Search Twitter”
- When tapped, the search engine suggestion allows users to change their search engine on a one-time basis

### String

- If an engine searches “the internet” (examples: Google, Bing, DuckDuckGo), the string should say:
    - Search with [providerName]
    - Examples: “Search with Google”, “Search with Bing”
- If an engine searches “the site itself” (examples: Twitter, Wikipedia, YouTube), the string should say:
    - Search [providerName]
    - Examples: “Search with Twitter”, “Search with Wikipedia”
- Underneath the label, there’s a caption. The string should say:
    - Search directly from the address bar

### Mockups

| On Autocomplete | After tapping “Search [providerName]” |
| --- | --- |
| ![Search Engine Autocomplete - Google@2x](https://user-images.githubusercontent.com/51007/96536672-1396cb80-12f1-11eb-9ef4-3530da169b70.png) | ![Search Engine Autocomplete - Amazon 4@2x](https://user-images.githubusercontent.com/51007/96536680-1b567000-12f1-11eb-9d5b-6963e09b696b.png) |

Ping @Verdi: I couldn’t get the provider name on the caption, because that would make the caption run too long. Would it be okay to omit it?

### Assets

- [search-with.svg](https://share.getcloudapp.com/o0um7prE/download/search-with.svg)

**Comment by mcarare** [comment]
@brampitoyo This seems like the same feature as selecting a search engine for one time use by selecting it from the "search engine" toggle list. Should we also change the hint on the address bar for that use case, to match the one described in this issue?

**Comment by mcarare** [comment]
@brampitoyo can you also provide the SVG asset as vector (XML)? Android Studio does not manage to convert it correctly. Thank you!

**Comment by brampitoyo** [comment]
Hi @mcarare, answering your questions inline:

1. Only one suggestion should be displayed
2. I didn’t know that a search engine is identified by a name rather than a URL. I would prefer the suggestion to be shown based on the URL. If we rely solely on name matching, people who type “bing.com” won’t get the suggestion. We want them to see the suggestion, even if the full URL is already entered.
3. For the time being, I don’t think we need to change the hint text for the other use case.
4. For the time being, let’s not worry about distinguishing between different types of search engine. We can just use the same string: “Search with [providerName]”

**Comment by mcarare** [comment]
> 
> If we rely solely on name matching, people who type “bing.com” won’t get the suggestion.

They will see it because the search suggestions appear as you type. But I will try to implement your design.

**Comment by mcarare** [comment]
@Bram On implementing this, I came across some cases when maybe using the autocomplete text not the typed one might not give us the expected result.

Let's say for example you visit `binge.com `. Then, when you type `b`, `bi`, `bin`, `bing` you would not get autocomplete to `bing.com` but to `binge.com` which would mean you will never get a search engine suggestion for Bing.

This would also apply to various scenarios when you will not autocomplete to a search engine except after typing many more letters to exclude autocompleting to a history or places URL.

**If this is expected** and we only want search engine suggestions for autocomplete I will go ahead with this implementation.

On the other hand, in the case described above, if using the entered text, by doing a partial matching we could suggest bing.
The only question would be when would we consider a match (from one letter, two or more, etc.)

What do you think?

**Comment by brampitoyo** [comment]
Hi @mcarare,

**TL;DR:** Yes, we only want search engine suggestions for autocomplete, but history items should always be ranked lower than hardcoded autocomplete and search engine entries. Right now, mobile ranks them higher.

I think the way mobile does autocomplete differs from desktop.

On desktop, even after I visited `binge.com` many times, typing `bi` will always show items in this order:
- Address bar: `bing.com`
- Item 1: `bing.com` (the autocomplete entry – hardcoded by the browser)
- Item 2: `Search with Bing` (the search engine entry – hardcoded by the browser)
- Item 3: `binge.com (the history entry)

Ideally, when I type `bi` on Fenix or Firefox for iOS, I also want to see things in the same order, except we don’t need `bing.com` repeated as item 1 (it’s already in the address bar):
- Address bar: `bing.com`
- Item 1: `Search with Bing`
- Item 2: `binge.com`

In other words: my history item should always appear lower than the hardcoded entries, but at the moment, it appears higher.

Maybe we should leave the fix as is, and then file another issue for tweaking the autocomplete recipe?

**Comment by mcarare** [comment]
@brampitoyo This would be another issue, the autocomplete order. The question is if we should roll this issue before or after reorganizing autocomplete.

As a note: I do not see `bing` at all in suggestions when I use Firefox desktop and type  `bi` (except in "this time search with"). The first entry and the autocomplete comes from history:

<img width="1170" alt="Screen Shot 2021-02-01 at 10 14 17 AM" src="https://user-images.githubusercontent.com/48995920/106431691-4d41ab80-6476-11eb-8cc1-43367bc16bc6.png">

**Comment by brampitoyo** [comment]
Hi @mcarare,

**TL;DR** – The behaviour you proposed was already perfect, and the bugs you’ve found was by design. There’s nothing to fix. Sorry for the trouble!

I’ve confirmed with Desktop Search UX that we want frecency to always “win” over any system-hardcoded value.

This means: in all the cases you demonstrated, the only way to get Bing to autocomplete and appear as a “Search with” item, is to type `bing.` (`bing` + `.`).

Otherwise, either `binge.com` or `bibnat.ro` will take over. And this is by design. If people are visiting a certain site a lot, we don’t want to block them by presenting our own hardcoded sites every single time. This would block them from getting to where they want to go.

So, the initial behaviour you proposed turned out to be correct. And ultimately, I was wrong for proposing that hardcoded entries should “win”. Sorry about this!

**Comment by mcarare** [comment]
Talked with @brampitoyo offline and we shall go ahead with using typed text instead of autocompleted text in order to avoid cases where we could not get a suggestion.

**Comment by LaurentiuApahideanSV** [comment]
Verified as fixed on Firefox Preview Nightly 210225 (Build #2015795331).
### Devices used:
- OnePlus 6T (Android 9)
- Google Pixel 3 XL (Android 9)
